### ISSUE TITLE
Use @types/estree for the parse method

### ISSUE THREAD

**Issue by aleclarson** [rollup/rollup#3800]
### Feature Use Case

Better types for `this.parse` in plugins.

### Feature Proposal

Re-export [`@types/estree`](https://www.npmjs.com/package/@types/estree) and return an `ESTree.Program` from the `parse` method, instead of the current `AcornNode` type.

**Comment by lukastaegert** [comment]
There was a lot of debate about types in general but in general it was deemed suboptimal to force those external types on everyone (though ESTree was not the biggest problem I admit). I vaguely remember there were also some inconsistencies with regard to how locations are handled in the ESTree types. Therefore I would recommend to do the type cast yourself.

**Comment by mayank99** [comment]
@lukastaegert Do you remember where these discussions were had? I would like to better understand the reasoning.

Also is there a recommended way of type casting? Simply doing `this.parse() as unknown as ESTree.Program` is not enough because all the nodes have `start`/`end` properties which is missing from estree nodes.

**Comment by lukastaegert** [comment]
Exactly. An `acorn.Node` is probably a more correct type, but is missing the specific information from the ESTree types. Maybe one can `&` the two types. I wouldn't easily find the old discussions in the issue, though.

**Comment by mayank99** [comment]
Looks like acorn's own types are not very detailed. `parse` just returns a `Node` which is missing properties like `body`. https://unpkg.com/browse/acorn@8.8.1/dist/acorn.d.ts

There's a definitely-typed package which makes use of estree types but it's deprecated so the types are likely outdated. https://unpkg.com/browse/@types/acorn@4.0.6/index.d.ts

I guess this is not an issue in rollup then and needs to be solved upstream. I found this github issue: https://github.com/acornjs/acorn/issues/946
